clusterman/autoscaler/pool_manager.py

```diff
@@ -438,6 +438,8 @@ def _choose_nodes_to_prune(
                     )
                     continue
 
+            logger.info(f"freezing {instance_id} for termination")
+            self.cluster_connector.freeze_agent(node_metadata.agent.agent_id)
```

Comment:
i'm still a little worried about racing against this taint now - in the time we get all nodes and iterate through them, we could be tainting a box that had something scheduled onto it (and ending up in the same situation where a box gets killed and was running something that doesn't tolerate interruptions)

Reply:
also, do we want to do anything about a node getting cordoned and clusterman crashing/getting killed/etc before terminating the cordoned box?

Reply:
For first question: It is part of main change which totally solve race condition. I am just applying change to production with small iteration. + I will check result after this change and I will solve race condition in case of necessity (because it is costy operation)

Reply:
For second question: there are two possibilities:
1) crash after draining_client.submit_instance_for_draining: We don't need to think about this case
2) crash before draining_client.submit_instance_for_draining: In this case clusterman will select same node as victim due to our sort logic (_prioritize_killable_nodes). But there is still a possibility (small) that other nodes (for ex: newly created nodes which have less tasks) are ahead of the line. We will have "cordoned (freezed)" node for short (hope) period. 

Thanks for mentioned it. 
I will dig a little bit.

Reply:
i'm not 100% sold that *just* cordoning termination candidates will permanently fix the issue we're seeing here, but i'm willing to try this out

Reply:
@nemacysts , I implemented "is_freezed" logic and added it to sort_key with the highest priority. So Clusterman will select this node for termination. What do you think? I'm still thinking about if I missed any edge cases.